### DISCUSSION TITLE
Theming and Module like functionality for OC websites

### DISCUSSION THREAD

**Discussion by nicpitsch** [OrchardCMS/OrchardCore#15099] [Q&A] [answered]
I feel with @FireEagle97 concerning [Discussion #15088](https://github.com/OrchardCMS/OrchardCore/discussions/15088). After searching for and consulting several sources - including [Dojo3](https://orcharddojo.net/orchard-training/dojo-course-3-the-full-orchard-core-tutorial), at least partial - I feel now a bit lost and a bit frustrated. Instead of fewer, there are more and more questions. 

I don't want to hijack the mentioned discussion, so I started this one. My hope is that my thoughts and conclusions so far do cause some guiding/clarifying answers from experienced users. 

My journey so far:

- One of the first thoughts I had about how to get into OC was how theming works.
- Most OrchardCore (OC) Themes seem to use [Liquid syntax](https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/).
- The one Theme with more than basic functionality I found is the [TheBlogTheme](https://github.com/OrchardCMS/OrchardCore/tree/main/src/OrchardCore.Themes/TheBlogTheme), using Liquid syntax. 
- But I don't want to learn Liquid, instead I want to use and enhance the existing C#/Razor knowledge using [Razor Views for Theming](https://docs.orchardcore.net/en/latest/docs/reference/modules/Liquid/#razor-helpers) instead. 
- So I thought how about recreate TheBlogTheme using Razor Views? Oh man TheBlogTheme has about 30 views - so difficult to understand and no clue how it would be made using Razor.
- Thinking more about the [TheBlogTheme](https://github.com/OrchardCMS/OrchardCore/tree/main/src/OrchardCore.Themes/TheBlogTheme) I have the suspicion it could maybe not be the best example for a Theme. Why, because it mixes styling with content/functionality. Imagine adding a Image Slider functionality to that Theme, we would end with maybe 50 views?
- Shouldn't a Blog, Gallery, Slider, FAQ etc. which provide functionality and data rather be distinct [Modules](https://docs.orchardcore.net/en/main/docs/getting-started/templates/#create-a-new-cms-module) like the [OC Modules](https://github.com/OrchardCMS/OrchardCore/tree/main/src/OrchardCore.Modules) instead of embeddeding such functionality within a Theme? 
- Would [Lombiq Modules](https://github.com/search?q=topic%253Aorchard-module+org%253ALombiq) or [Etch Modules](https://github.com/orgs/EtchUK/repositories?q=module) or [Recipes Module for Orchard Core](https://github.com/agriffard/Recipes.OrchardCore) be a starting point?
- At the time I'm going through the [Orchard-Training-Demo-Module](https://github.com/Lombiq/Orchard-Training-Demo-Module). The code is very well documented and also the [StartLearningHere](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/StartLearningHere.md) is useful. 
- Now that I'm there - and a bit lost considering the complexity of the *Orchard-Training-Demo-Module* - I'm wondering if I first should have more knowledge about [ASP.NET Core (MVC)](https://learn.microsoft.com/en-us/aspnet/core/mvc/overview?view=aspnetcore-8.0).

This isn't a complaint - on the contrary, OC documentation is very fine and there are a lot of resources around and the GitHub Issues and Discussions here are well supported by a lot of helpful people. But I feel like I have not found the right way to get started with OC building websites, with the aim of include more than just basic functionality.

If I would know it better I would give to myself following recommendations (but I could also be completely wrong :confused:):

- Take e. g. the [TheBootstrapTheme](https://github.com/agriffard/TheBootstrapTheme) as an example for a Razor based Theme.
- Don't implement much more functionality than what is there (primarily Menu/Navigation).
- For simple page content and similar pieces use the functionality the OC Admin UI offers through the usage of Content Types and Content Parts and so on. Put that into Recipes for reuse.
- For further functionality like Blog, Gallery, Slider, FAQ etc. learn to program Modules. This is where you can use your C# ASP.NET Core (MVC) knowledge. If I only could tell you that there are a lot of such Modules already around - ready to use as a starting point.

Which recommendations would you give?

**Comment by FireEagle97** [comment]
Since we are on the same Boat, I will share my thoughts with you. I don't know liquid either and not planning to learn it. I'm planning to find a way to modify a theme other than the default one like the blog theme. Do I need to add theBlog theme to the Theme dir and modify it directly if not I will try to import my theme. Also, I'm planning to find a way to modify the Admin theme and build on it.

**Comment by FireEagle97** [comment]
I just found this YouTube video about creating a new theme. I thought it could help you. You need to download the dependencies from Nuget instead of having them in the same folders and you don't need to use Liquid.
[https://www.youtube.com/watch?v=wtAIgn4gYXc&ab_channel=Lombiq](url)

**Reply by nicpitsch** [reply]
Thanks. I prefer the [TheBootstrapTheme](https://github.com/agriffard/TheBootstrapTheme) which is a Razor based Theme, as a starting point for Theme development. If I need more control I could add more functionality using further Razor Views.

**Comment by nicpitsch** [comment]
> If I would know it better I would give to myself following recommendations (but I could also be completely wrong 😕):
> 
>     * For further functionality like Blog, Gallery, Slider, FAQ etc. learn to program Modules. This is where you can use your C# ASP.NET Core (MVC) knowledge. If I only could tell you that there are a lot of such Modules already around - ready to use as a starting point.

I'm wondering whether I'm right with the above assumption, that further functionality like Blog, Gallery, Slider, FAQ should be implemented with Modules and not within a Theme.

**Comment by Piedone** [comment, marked as answer]
I'm not sure I have unpacked everything here, but let me try to help :).

- I highly recommend you work through the mentioned [Dojo Course 3](https://orcharddojo.net/orchard-training/dojo-course-3-the-full-orchard-core-tutorial) fully, not just partially. Its tutorial videos build on each other, so you'll be able to understand more and more concepts of Orchard Core, and you'll begin to see what the different approaches are good for. It also explains how and when you'd use built-in OC features, build something from the admin vs from code, and when to create a module or a theme.
- Technically, modules and themes are very similar, and you can add any kind of functionality to both of them. Semantically, a theme is to provide primarily styling and display logic (i.e. Liquid or Razor templates, CSS, JS, graphics), and a module to provide functionality independent of themes. You can also override any part of the OC layout (the HTML output going to the browser) from the current site or admin theme (the theme always wins), while a module can be overridden by a theme.
- Content type and part settings, i.e. content definitions are always (well, almost, don't worry about it for now) stored in the OC database. If you create/configure content types from the dashboard, then that's all to it. However, you can initialize (and later update) content definitions from a recipe or a migration class too for reuse.

**Reply by nicpitsch** [reply]
Thanks for the valuable guidance!
